From f5cdeeb165d988b33af12edc437e60d68fb95a8e Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:53:49 -0400
Subject: [fs] ext4: don't count free clusters from a corrupt block group

Message-id: <1414058125-4183-76-git-send-email-lczerner@redhat.com>
Patchwork-id: 98155
O-Subject: [RHEL 7.1 PATCH v2 075/171] ext4: don't count free clusters from a corrupt block group
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: "Darrick J. Wong" <darrick.wong@oracle.com>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit 2746f7a17062d3526116f7ae7f91d88b19c2464e

 A bg that's been flagged "corrupt" by definition has no free blocks,
 so that the allocator won't be tempted to use the damaged bg.
 Therefore, we shouldn't count the clusters in the damaged group when
 calculating free counts.

 Signed-off-by: Darrick J. Wong <darrick.wong@oracle.com>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Reviewed-by: Zheng Liu <wenqing.lz@taobao.com>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/balloc.c b/fs/ext4/balloc.c
index dc5d572..6ea7b14 100644
--- a/fs/ext4/balloc.c
+++ b/fs/ext4/balloc.c
@@ -640,6 +640,7 @@ ext4_fsblk_t ext4_count_free_clusters(struct super_block *sb)
  struct ext4_group_desc *gdp;
  ext4_group_t i;
  ext4_group_t ngroups = ext4_get_groups_count(sb);
+ struct ext4_group_info *grp;
 #ifdef EXT4FS_DEBUG
  struct ext4_super_block *es;
  ext4_fsblk_t bitmap_count;
@@ -655,7 +656,11 @@ ext4_fsblk_t ext4_count_free_clusters(struct super_block *sb)
   gdp = ext4_get_group_desc(sb, i, NULL);
   if (!gdp)
    continue;
-  desc_count += ext4_free_group_clusters(sb, gdp);
+  grp = NULL;
+  if (EXT4_SB(sb)->s_group_info)
+   grp = ext4_get_group_info(sb, i);
+  if (!grp || !EXT4_MB_GRP_BBITMAP_CORRUPT(grp))
+   desc_count += ext4_free_group_clusters(sb, gdp);
   brelse(bitmap_bh);
   bitmap_bh = ext4_read_block_bitmap(sb, i);
   if (bitmap_bh == NULL)
@@ -679,7 +684,11 @@ ext4_fsblk_t ext4_count_free_clusters(struct super_block *sb)
   gdp = ext4_get_group_desc(sb, i, NULL);
   if (!gdp)
    continue;
-  desc_count += ext4_free_group_clusters(sb, gdp);
+  grp = NULL;
+  if (EXT4_SB(sb)->s_group_info)
+   grp = ext4_get_group_info(sb, i);
+  if (!grp || !EXT4_MB_GRP_BBITMAP_CORRUPT(grp))
+   desc_count += ext4_free_group_clusters(sb, gdp);
  }
 
  return desc_count;
-- 
1.7.1